R/Spatial.R

```diff
@@ -0,0 +1,180 @@
+#' @export
+RUVg_Housekeeping_Normalization <- function(seuratObj, assay = "RNA", k = 1){
+  housekeeping_genes <- c("ABCF1", "ACTB", "ALAS1", "B2M", "CLTC", "G6PD", "GAPDH", 
+                          "GUSB", "HPRT1", "LDHA", "PGK1", "POLR1B", "POLR2A",
+                          "RPL19", "RPLP0", "SDHA", "TBP", "TUBB")
+  counts <- as.matrix(seuratObj[[assay]]@counts)
+  rounded_counts <- round(as.matrix(counts))
+  RUVg <- RUVSeq::RUVg(rounded_counts,
+                       housekeeping_genes[housekeeping_genes %in% rownames(seuratObj)], 
+                       k = k)
+  seuratObj[['RUVg']] <- Seurat::CreateAssayObject(as(RUVg$normalizedCounts, "dgCMatrix"))
+  return(seuratObj)
+}
+
+#' @title NanoString_Housekeeping_Normalization
+#'
+#' @description a normalization method that uses the housekeeping genes to generate size factors for normalization.
+#' @param seuratObj The Seurat object that holds the data to be normalized.
+#' @param assay The assay holding the raw data - the counts slot will be assumed to hold raw data.
+#' @return Returns a Seurat object containing normalized counts stored in the counts slot of the "Housekeeping" assay. 
+#' @export
+
+NanoString_Housekeeping_Normalization <- function(seuratObj, assay = "RNA"){
+  
+  #Normalization Steps from Nanostring 
+  #1.  Calculate the geometric mean of the selected housekeeping genes for each lane.
+  #2. Calculate the arithmetic mean of these geometric means for all sample lanes.
+  #3. Divide this arithmetic mean by the geometric mean of each lane to generate a lane-specific normalization factor.
+  #4. Multiply the counts for every gene by its lane-specific normalization factor.
+  
+  counts <- as.matrix(seuratObj[[assay]]@counts) 
+  
+  housekeeping_genes <- c("ABCF1", "ACTB", "ALAS1", "B2M", "CLTC", "G6PD", "GAPDH", 
+                          "GUSB", "HPRT1", "LDHA", "PGK1", "POLR1B", "POLR2A",
+                          "RPL19", "RPLP0", "SDHA", "TBP", "TUBB")
+  housekeeping_counts <- counts[rownames(counts) %in% housekeeping_genes, ]
+  
+  #Ensure no downstream division by zero errors
+  if(any(compositions::geometricmeanRow(as.matrix(t(housekeeping_counts)) == 0))){
+    stop("Error: Geometric mean of housekeeping counts for one or more samples is zero. One of the housekeeping genes likely has a value of zero. Either add a pseudocount (if appropriate) or choose a different normalization")
+  }
+  
+  #steps 1-3 of the normalization are computed here
+  Sample_Normalization_Factors <- mean(compositions::geometricmeanRow(as.matrix(t(housekeeping_counts)))) / compositions::geometricmeanRow(as.matrix(t(housekeeping_counts)))
+  
+  #step 4 of the normalization is computed here
+  hk_normalized_counts <- housekeeping_counts %*% diag(Sample_Normalization_Factors)
+  colnames(hk_normalized_counts) <- colnames(counts)
+  seuratObj[["Housekeeping"]] <- Seurat::CreateAssayObject(counts = as(hk_normalized_counts, "dgCMatrix"))
```

Comment:
same comment as above about name of target and also whether it's actually a good idea to store into another assay